File: 24. BitmaskDP_SubsetSum.py
# Topic: Simple bitmask-based DP for subset sum using Python's integer bitmask

def subset_sum_bitmask(arr, target):
    bs = 1
    for x in arr:
        bs |= (bs << x)
    return (bs >> target) & 1 == 1

if __name__ == '__main__':
    print('Subset sum 9 exists:', subset_sum_bitmask([3,34,4,12,5,2],9))

# Time Complexity: O(n * target / wordsize)
# Space Complexity: O(target)

File: 24. BitmaskDP_SubsetSum.cpp
// Topic: Simple bitmask-based DP for subset sum using bitset (efficient for small sums)
#include <bits/stdc++.h>
using namespace std;

bool subsetSumBitset(const vector<int>& arr, int target)
{
    bitset<10001> bs;
    bs[0] = 1;
    for (int x : arr) bs |= (bs << x);
    return bs[target];
}

int main()
{
    vector<int> a = {3, 34, 4, 12, 5, 2};
    cout << "Subset sum 9 exists: " << (subsetSumBitset(a, 9) ? "Yes" : "No") << '\n';
    return 0;
}

// Time Complexity: O(n * target / wordsize)
// Space Complexity: O(target)
